### ISSUE TITLE
No way to set maxSessions per browser type in Selenium Grid 4

### ISSUE THREAD

**Issue by hnestorova** [SeleniumHQ/selenium#8974]
## 🐛 Bug Report

In the previous grid version, we could specify maxSessions and instances like this:
browser browserName="chrome",platform=WINDOWS,maxInstances=1 -maxSession 1

With Selenium Grid 4, up to now, I have not found a way to modify the default values which will be 1 for IE, 2 for Chrome and 2 for Edge.

## To Reproduce
add a node as:
java -jar selenium-server-4.0.0-alpha-7.jar node --detect-drivers true --publish-events tcp://HOST_IP:PORT --subscribe-events tcp://HOST_IP:PORT

## Expected behavior

Able to specify how many sessions per browser type are desired to run on a single node at the same time

## Test script or set of commands reproducing this issue

N/A

## Environment

OS: Windows 7
Browser: IE, Chrome and Edge
Browser version: IE 11.0.9600.19356
Browser Driver version: 3.150.1
Selenium Grid version : Grid 4 (4.0.0-alpha-7)

**Comment by hnestorova** [comment]
To further explain, due to this, if I run tests with parallel classes on Chrome, the first two classes start being run on the same VM and they fail.

**Comment by pujagani** [comment]
Certain browsers like IE do not support multiple concurrent session. There are issues related to browser isolation. Hence it defaults to 1. (Thank you Simon and Jim for helping me understand this). 
For browsers like Chrome and Mozilla which allow multiple concurrent sessions, the "--max-sessions" flag can be used to provide the max concurrent sessions value. However, it will be allocated based on the available processors at the moment. The default values that are mentioned are also based on available processors for Chrome and Mozilla. 
If enough processors are available, then the max-sessions value is considered else it will stick to the default values.

**Comment by pujagani** [comment]
Can you help me understand the issue with running two chrome tests in parallel better? Would appreciate the error, tests or code sample. Please mention the set-up as well. Are you running the Grid in fully distributed mode or using Hub and node mode? If you think the issue is not related to setting max-sessions, then I would suggest please create a separate issue. It will easier to address it that way. Thank you!

**Comment by hnestorova** [comment]
Hi, sorry for the late reply. The --max-sessions flag solved the problem with starting multiple Chrome browsers on the same VM which broke our tests. 

However, with our current setup, when we have a single node, --max-sessions 1, thread-count=2, 2 Java test classes, parallel set to classes in the testng xml we encounter a SessionNotCreatedException . Note that the problem seems to be that we have two Java test classes,parallel set to classes, thread-count set to 2 but only one available node

Here is our setup:

- Hub
start C:/"Program Files"/Java/jdk-11.0.2/bin/java -jar selenium-server-4.0.0-alpha-7.jar hub --port 4452 --publish-events tcp://HUB_IP:4455 --subscribe-events tcp://HUB_IP:4456

- Node
start C:/"Program Files"/Java/jdk-11.0.2/bin/java -jar selenium-server-4.0.0-alpha-7.jar node --detect-drivers true --publish-events tcp://HUB_IP:4455 --subscribe-events tcp://HUB_IP:4456 --max-sessions 1
 
- TestNG xml content
`<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd">
<suite name="Test Classes" parallel="classes" thread-count="2">
  	<parameter name="baseUrl" value="PRIVATE"/>
	<parameter name="db" value="PRIVATE"/>
	<parameter name="userGroup" value="PRIVATE"/>
	<parameter name="userName" value="PRIVATE"/>
	<parameter name="password" value="PRIVATE"/>
	<parameter name="adminGroup" value="PRIVATE"/>
	<parameter name="admin" value="PRIVATE"/>
	<parameter name="adminPass" value="PRIVATE"/>
	<parameter name="serverURL" value="http:/PRIVATE:4452/wd/hub"/>
	<parameter name="browserName" value="chrome"/>
	
	<test verbose="1" name="PRIVATE" annotations="JDK">
		<classes>
		<class name="PRIVATE.TestClass1"/>
                <class name="PRIVATE.TestClass2"/>
      </classes>
	</test>
    <listeners>
        <listener class-name="PRIVATE.ScreenshotListener"/>
    </listeners>
</suite>`

- Java Test Classes content
**TestClass1**
@Test(groups = { "all" }, description = "Here goes the description.")
	public void testCase2() {
		System.out.println("Running Case 1");
	}
**TestClass2**
@Test(groups = { "all" }, description = "Here goes the description.")
	public void testCase2() {
		System.out.println("Running Case 2");
	}
- Exception when we have a single node, --max-sessions 1, thread-count=2, 2 Java test classes, parallel set to classes in the testng xml
`[RemoteTestNG] detected TestNG version 7.3.0
SLF4J: Class path contains multiple SLF4J providers.
SLF4J: Found provider [org.slf4j.jul.JULServiceProvider@b6fcb2b]
SLF4J: Found provider [org.slf4j.log4j12.Log4j12ServiceProvider@3c200c00]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual provider is of type [org.slf4j.jul.JULServiceProvider@b6fcb2b]
Jan 04, 2021 4:25:38 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: OSS
2021-01-04 16:25:50 running 'testCase1' on node 'PRIVATE'
Running Case 1

===============================================
Test Classes
Total tests run: 2, Passes: 1, Failures: 0, Skips: 1
Configuration Failures: 1, Skips: 3
===============================================

org.openqa.selenium.SessionNotCreatedException: Unable to find provider for session: Capabilities {browserName: chrome, goog:chromeOptions: {args: [disable-infobars], extensions: [Q3IyNAMAAAAdBQAAEqwECqYCMII...], prefs: {credentials_enable_service: false, download.default_directory: \\PRIVATE\E..., profile.password_manager_enabled: false}}}
Build info: version: '4.0.0-alpha-7', revision: 'Unknown'
System info: host: 'PRIVATE', ip: 'PRIVATE', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '11.0.2'
Driver info: driver.version: unknown
Build info: version: '4.0.0-alpha-7', revision: 'Unknown'
System info: host: 'PRIVATE', ip: 'PRIVATE', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '11.0.2'
Driver info: driver.version: RemoteWebDriver
remote stacktrace: org.openqa.selenium.SessionNotCreatedException: Unable to find provider for session: Capabilities {browserName: chrome, goog:chromeOptions: {args: [disable-infobars], extensions: [Q3IyNAMAAAAdBQAAEqwECqYCMII...], prefs: {credentials_enable_service: false, download.default_directory: \\PRIVATE\E..., profile.password_manager_enabled: false}}}
Build info: version: '4.0.0-alpha-7', revision: 'Unknown'
System info: host: 'PRIVATE', ip: 'PRIVATE', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '11.0.2'
Driver info: driver.version: unknown
	at org.openqa.selenium.grid.distributor.Distributor.newSession(Distributor.java:174)
	at org.openqa.selenium.grid.distributor.Distributor.lambda$new$0(Distributor.java:146)
	at org.openqa.selenium.remote.http.Route$TemplatizedRoute.handle(Route.java:183)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:67)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:327)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:67)
	at org.openqa.selenium.grid.distributor.Distributor.execute(Distributor.java:313)
	at org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:86)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:327)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:67)
	at org.openqa.selenium.grid.router.Router.execute(Router.java:85)
	at org.openqa.selenium.grid.web.CheckOriginHeader.lambda$apply$0(CheckOriginHeader.java:66)
	at org.openqa.selenium.grid.web.CheckContentTypeHeader.lambda$apply$0(CheckContentTypeHeader.java:70)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:327)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:67)
	at org.openqa.selenium.remote.http.Route$NestedRoute.handle(Route.java:261)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:67)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:327)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:67)
	at org.openqa.selenium.grid.server.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:30)
	at org.openqa.selenium.grid.server.WrapExceptions.lambda$apply$0(WrapExceptions.java:36)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
	at org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:46)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.openqa.selenium.grid.distributor.RetrySessionRequestException: Unable to find provider for session: Capabilities {browserName: chrome, goog:chromeOptions: {args: [disable-infobars], extensions: [Q3IyNAMAAAAdBQAAEqwECqYCMII...], prefs: {credentials_enable_service: false, download.default_directory: \\PRIVATE\E..., profile.password_manager_enabled: false}}}
Build info: version: '4.0.0-alpha-7', revision: 'Unknown'
System info: host: 'PRIVATE', ip: 'PRIVATE', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '11.0.2'
Driver info: driver.version: unknown
	at org.openqa.selenium.grid.distributor.Distributor.createNewSessionResponse(Distributor.java:261)
	at org.openqa.selenium.grid.distributor.Distributor.newSession(Distributor.java:168)
	... 29 more

	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.openqa.selenium.remote.W3CHandshakeResponse.lambda$errorHandler$0(W3CHandshakeResponse.java:62)
	at org.openqa.selenium.remote.HandshakeResponse.lambda$getResponseFunction$0(HandshakeResponse.java:30)
	at org.openqa.selenium.remote.ProtocolHandshake.lambda$createSession$1(ProtocolHandshake.java:126)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:128)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:75)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:139)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:501)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:207)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:127)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:140)
	at com.bravogroup.cognity.web.selenium.tests.BaseTest.getRemoteDriver(BaseTest.java:504)
	at com.bravogroup.cognity.web.selenium.tests.BaseTest.createChromeDriver(BaseTest.java:561)
	at com.bravogroup.cognity.web.selenium.tests.BaseTest.init(BaseTest.java:431)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:61)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:366)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:320)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:176)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
`

**Comment by hnestorova** [comment]
@diemol  answer was provided but i am unable to remove the label.

**Comment by pujagani** [comment]
Thank you for the reply and for providing the details. Can you please help me understand as to why is the --max-sessions value set to 1?  I think due to that the 2nd request for a Chrome session is getting rejected. If the max-sessions is set to 2 and the same number of processors are available then it should work. The current version in the trunk queues the incoming request, in cases where the capability exists in the node but the Node does not have the capacity (Example: Here 1st test has occupied the Chrome slot on the Node and hence 2nd request for the same slot get rejected), the 2nd request will not get rejected if the capability exists. It will be added back to the queue and retried.

**Comment by hnestorova** [comment]
Hi @pujagani 

In our setup which we have used for more than 9 years already, we want to have only one test running per Virtual machine. The reason for this is that otherwise, browsers loose focus and etc which makes our tests not stable.

To achieve this with the previous grid version, we had our nodes configured like this for when Chrome was used:
java -Dwebdriver.ie.driver="chromedriver.exe" -jar selenium-server-standalone.jar -role node -hub http://PRIVATE:4447/grid/register -browser browserName="chrome",platform=WINDOWS,maxInstances=1 -maxSession 1

Can you tell me with the new Selenium Grid 4, how I can accomplish the same setup - each Virtual Machine in the grid to have only a single browser used by the tests at a given moment?

**Comment by pujagani** [comment]
Thank you for helping me understand the setup. Now I have a better context of the situation and the setup. As far I understand, the previous grid version had a queue for new session request and hence the tests could run in the setup mentioned. The alpha-7 release does not have the queue fully integrated into Grid 4. As I mentioned earlier, hence the request gets rejected when max-session is set to 1, because it does not queue the 2nd request and simply rejects it with the error message. The current version in the trunk has the queue integrated into Grid 4. It will be released as part of the next release which is the first beta. With that and the setup you have in place, it should be able to handle the two tests.

**Comment by pujagani** [comment]
Thanks to the changes landed by Diego recently, https://github.com/SeleniumHQ/selenium/blob/trunk/java/server/src/org/openqa/selenium/grid/node/httpd/NodeFlags.java. The node can be configured with the desired webdriver configured for a browser and setting the max-session to 1, will ensure that only one browser will run at a time.

**Comment by pujagani** [comment]
With the current trunk or the upcoming version, if you run `start C:/"Program Files"/Java/jdk-11.0.2/bin/java -jar selenium-server-4.0.0-alpha-7.jar node --detect-drivers true -I "chrome" --publish-events tcp://HUB_IP:4455 --subscribe-events tcp://HUB_IP:4456 --max-sessions 1` then you should be able to achieve the required setup.

**Comment by hnestorova** [comment]
Hi @pujagani 
Thanks for your answer. So, the fix will be available in the next alpha version? And, if so, do you know when it is expected to be released? Or did you mean that if we use the set up from your last comment where browser type (chrome) is explicitly specified, it should work with the current version as well?

Update: I just tried it and it opened two chrome browsers at the same time on the same machine, so i guess it will be added to the next release?

**Comment by diemol** [comment]
I implemented this through the `driver-configuration` option added a some days ago.

In short, you can provide a configuration file like this one:
```toml
[node]
detect-drivers = false
[[node.driver-configuration]]
name = "Firefox Nightly"
stereotype = "{\"browserName\": \"firefox\", \"browserVersion\": \"86\", \"moz:firefoxOptions\": {\"binary\": \"/Applications/Firefox Nightly.app/Contents/MacOS/firefox-bin\"}}"
max-sessions = 4
[[node.driver-configuration]]
name = "Chrome Beta"
stereotype = "{\"browserName\": \"chrome\", \"browserVersion\": \"88\", \"goog:chromeOptions\": {\"binary\": \"/Applications/Google Chrome Beta.app/Contents/MacOS/Google Chrome Beta\"}}"
max-sessions = 4
[[node.driver-configuration]]
name = "Firefox"
stereotype = "{\"browserName\": \"firefox\", \"browserVersion\": \"84\"}"
max-sessions = 4
[[node.driver-configuration]]
name = "Chrome"
stereotype = "{\"browserName\": \"chrome\", \"browserVersion\": \"87\"}"
max-sessions = 4
[[node.driver-configuration]]
name = "Internet Explorer"
stereotype = "{\"browserName\": \"internet explorer\"}"
max-sessions = 1
```

The `max-sessions` are respected unless they are higher than the amount of CPUs available, and for some cases, like Safari and IE, it will always use 1.

If you'd like to use a jar that includes this change, please check https://github.com/SeleniumHQ/docker-selenium/tree/beta-jars in a few hours. Over there we publish pre-released jars that will be used in the docker-selenium project, but they can also be used as they are.

**Comment by hnestorova** [comment]
Hi
@diemol , thanks, but I am not using Docker. My setup as pointed already is
Hub
start C:/"Program Files"/Java/jdk-11.0.2/bin/java -jar selenium-server-4.0.0-alpha-7.jar hub --port 4452 --publish-events tcp://HUB_IP:4455 --subscribe-events tcp://HUB_IP:4456

Node
start C:/"Program Files"/Java/jdk-11.0.2/bin/java -jar selenium-server-4.0.0-alpha-7.jar node --detect-drivers true --publish-events tcp://HUB_IP:4455 --subscribe-events tcp://HUB_IP:4456 --max-sessions 1

**Comment by diemol** [comment]
@hnestorova, you do not need to use docker for this, what I meant is, if you want to grab a jar recently built with this feature, you can grab it from the link I placed above.

**Comment by hnestorova** [comment]
@diemol thanks , I tried it but got another issue - https://github.com/SeleniumHQ/selenium/issues/9079

**Comment by mikej411** [comment]
I have been using a set of virtual machines for years executing 10 max sessions of Chrome on older versions of Selenium Server. I spun up Server 4 beta and it’s telling me the limit is 2 sessions. Does this mean I’ll only ever be able to execute 2 chrome windows at time? If so, is this going to change so I can set a higher value? It does not seem very smart to set a hard cap/limit based on processors. You should leave it up the the tester

**Comment by mikej411** [comment]
I am running Server 4 Beta 1, and I can not set my max sessions above 2. From what I read, Server 4 is limiting the max sessions based on the servers processor count. Is this just temporary until the full release comes out? If not, why is it limiting? I have been executing 10 tests in parallel on each server with no issues, and now I have to limit that to only 2 tests if I want to upgrade? Is there any workaround I can apply to increase max sessions in Grid 4?

I try this...
java -jar selenium-server-4.0.0-beta-1.jar node --detect-drivers true --max-sessions 10
But the Max Concurrency stays at 2

**Comment by mikej411** [comment]
> However, it will be allocated based on the available processors at the moment

What does "at the moment" mean? Will this cap/limit be lifted in a future release, and if so, when is the ETA for that? I want to upgrade, but the limit based on processor count is stopping me. I dont think this limit makes very much sense, as Im able to execute 10 tests in parallel on all of my servers with Grid Server 3, and I only have 2 processors on each

**Comment by SnakeULTIMATE** [comment]
@mikej411 @hnestorova Good news. I found the answer
Tested on beta 1
java -XX:ActiveProcessorCount=12 -jar selenium4.jar standalone --max-sessions 12

**Comment by mikej411** [comment]
@SnakeULTIMATE Thank you! What do I replace the **XX** with?

**Comment by SnakeULTIMATE** [comment]
@mikej411 This is a parameter XX:ActiveProcessorCount=12. Just replace 12 on your preferred max sessions

**Comment by IssacH** [comment]
Hi @SnakeULTIMATE, @mikemelia @pujagani , I'd like to know how to set max sessions in docker yml file. Is there a documentation? Or would you please guide me? Thanks very much. BTW, I use the latest docker image for selenium hub and node

**Comment by SnakeULTIMATE** [comment]
@IssacH Sorry. I don't khow. I found the answer here - https://github.com/SeleniumHQ/selenium/issues/8730

**Comment by pujagani** [comment]
Hello,
The convention followed is  "SE_`component-name`_`option-name with underscores`". 

So for node max sessions, it would be: SE_NODE_MAX_SESSIONS=3
To see all options for the Grid component refer: https://www.selenium.dev/documentation/en/grid/grid_4/configuring_components/config_help/

This should be passed in the ENVIRONMENT section of the docker-compose file.

Sharing a quick example below : 

``` bash 
version: '3.7'

services:
  hub4:
    image: selenium/hub:4.0.0-beta-2-20210317
    ports:
      - "4442:4442"
      - "4443:4443"
      - "4444:4444"

  chromenode:
    image: selenium/node-chrome:4.0.0-beta-2-20210317
    volumes:
      - /dev/shm:/dev/shm
    depends_on:
      - hub4
    environment:
      - SE_EVENT_BUS_HOST=hub4
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SE_NODE_SESSION_TIMEOUT=500
      - SE_NODE_MAX_SESSIONS=3
    ports:
      - "6900:5900"


```

**Comment by IssacH** [comment]
Hi @pujagani,

Your reply is really helpful and now it works! Thanks very much!

**Comment by sqli0189** [comment]
> Hi @SnakeULTIMATE, @mikemelia @pujagani , I'd like to know how to set max sessions in docker yml file. Is there a documentation? Or would you please guide me? Thanks very much. BTW, I use the latest docker image for selenium hub and node

Hi @IssacH, I followed @SnakeULTIMATE comments and successfully increase the max sessions, my selenium grid 4 is deployed on the K8S, in my yaml file, I add below environment variables. 

```
env:
    - name: JAVA_OPTS
       value: "-XX:ActiveProcessorCount=5"
    - name: SE_NODE_MAX_SESSIONS
       value: "5"
```

So for Docker, I guess you can set active processor count by adding below environment variable when you launch the selenium node container:

```
docker run --rm -e JAVA_OPTS='-XX:ActiveProcessorCount=5' .....
```

Good luck!
